encoding/dot/decode.go

```diff
@@ -0,0 +1,223 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package dot
+
+import (
+	"fmt"
+
+	"github.com/gonum/graph"
+	"github.com/gonum/graph/encoding/dot/dotparser/ast"
+)
+
+// Builder is a graph that can have user-defined nodes and edges added.
+type Builder interface {
+	graph.Graph
+	graph.Builder
+	// NewNode adds a new node with a unique node ID to the graph.
+	NewNode() graph.Node
+	// NewEdge adds a new edge from the source to the destination node to the
+	// graph, or returns the existing edge if already present.
+	NewEdge(from, to graph.Node) graph.Edge
+}
+
+// UnmarshalerAttr is the interface implemented by objects that can unmarshal a
+// DOT attribute description of themselves.
+type UnmarshalerAttr interface {
+	// UnmarshalDOTAttr decodes a single DOT attribute.
+	UnmarshalDOTAttr(attr Attribute) error
+}
+
+// Copy copies the nodes and edges from the Graphviz AST source graph to the
+// destination graph. Edge direction is maintained if present.
+func Copy(dst Builder, src *ast.Graph) (err error) {
```

Comment:
I agree that using something along the lines of `Unmarshal(src []byte, v interface{}) error` would be cleaner, as then, the DOT AST details would not have to be explosed to the user.

I tried to implement this, but got stuck on a stumbling block.

It works well for when `v` is of type `dot.Builder`. However, using `*[]dot.Builder` doesn't seem to work, as the slice provided by the user would be empty, and the `Unmarshal` function would then append the decoded graphs to the slice. This would work, except, the `Unmarshal` function needs a user provided graph builder to be able to invoke the `NewNode` and `NewEdge` functions. Otherwise there is no way for the `Unmarshal` function to create user-defined types of nodes and edges.

Ideally. the user would know how many graphs are contained within a given file, and may then pass that number of `dot.Builder`s in a slice. If this was the case, the type of `v` could be `[]dot.Builder` instead of `*[]dot.Builder`, as the `Unmarshal` function would never append any new builders by itself, and thus does not need a pointer to slice.

Another approach would be for the user to pass one builder in the `*[]dot.Builder` slice and introduce some sort of `Dup` method on builders, to allow `Unmarshal` to create the number of builders needed from a single user-defined graph builder. Personally, I don't like this approach, as it makes the interface of dot.Builder even larger, and the implementation of `Dup` may at times be non-trivial depending on the representation of the user-defined graph.

Any ideas? What am I missing. I really hope we may resolve this final issue to get a clean API to work with.